examples/vision/nerf.py

```diff
@@ -0,0 +1,676 @@
+            rgb, _ = render_rgb_depth(
+                model=self.nerf_model, rays_flat=rays_flat, t_vals=t_vals, rand=True
+            )
+            loss = self.loss(images, rgb)
+
+        # Get the trainable variables.
+        trainable_variables = self.nerf_model.trainable_variables
+
+        # Get the gradeints of the trainiable variables with respect to the loss.
+        gradients = tape.gradient(loss, trainable_variables)
+
+        # Apply the grads and optimize the model.
+        self.optimizer.apply_gradients(zip(gradients, trainable_variables))
+
+        # Get the PSNR of the reconstructed images and the source images.
+        psnr = tf.image.psnr(images, rgb, max_val=1.0)
+
+        # return the loss and the psnr value
+        return {"loss": loss, "psnr": psnr}
+
+
+test_imgs, test_rays = next(iter(train_ds))
+test_rays_flat, test_t_vals = test_rays
+
+loss_list = []
+
+
+class CustomCallback(keras.callbacks.Callback):
+    def on_epoch_end(self, epoch, logs=None):
+        loss = logs["loss"]
+        loss_list.append(loss)
+        test_recons_images, depth_maps = render_rgb_depth(
+            model=self.model.nerf_model,
+            rays_flat=test_rays_flat,
+            t_vals=test_t_vals,
+            rand=True,
+            train=False,
+        )
+
+        # Plot the rgb, depth and the loss plot.
+        fig, ax = plt.subplots(nrows=1, ncols=3, figsize=(20, 5))
+        ax[0].imshow(keras.preprocessing.image.array_to_img(test_recons_images[0]))
+        ax[0].set_title(f"Predicted Image: {epoch:03d}")
+
+        ax[1].imshow(keras.preprocessing.image.array_to_img(depth_maps[0, ..., None]))
+        ax[1].set_title(f"Depth Map: {epoch:03d}")
+
+        ax[2].plot(loss_list)
+        ax[2].set_xticks(np.arange(0, EPOCHS + 1, 5.0))
+        ax[2].set_title(f"Loss Plot: {epoch:03d}")
+
+        fig.savefig(f"images/{epoch:03d}.png")
+        plt.close()
+
+
+num_pos = H * W * NUM_SAMPLES
+nerf_model = get_nerf_model(num_layers=8, num_pos=num_pos)
+
+model = NeRF(nerf_model)
+model.compile(optimizer=keras.optimizers.Adam(), loss=keras.losses.MeanSquaredError())
+
+# Create a directory to save the images during training.
+if not os.path.exists("images"):
+    os.makedirs("images")
+
+model.fit(
+    train_ds,
+    batch_size=BATCH_SIZE,
+    epochs=EPOCHS,
+    callbacks=[CustomCallback()],
+    steps_per_epoch=num_images // BATCH_SIZE,
+)
+
+
+def create_gif(path_to_images, name_gif):
+    filenames = glob.glob(path_to_images)
+    filenames = sorted(filenames)
+    images = []
+    for filename in tqdm(filenames):
+        images.append(imageio.imread(filename))
+    kargs = {"duration": 0.25}
+    imageio.mimsave(name_gif, images, "GIF", **kargs)
+
+
+create_gif("images/*.png", "training.gif")
+
+"""
+## Visualize the training step
+
+Here we see the training step. With decreasing loss, the rendered
+image and the depth maps are getting better.
+"""
+
+Image(open("training.gif", "rb").read())
```

Comment:
This is jupyter-specific, pefer using matplotlib or similar to display the image so that the code works in a terminal